### PR TITLE
Let clients declare Gradle distribution for imported projects

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @author Donat Csikos
+ * @since 3.0
+ */
+public abstract class GradleDistribution {
+
+    /**
+     * The available Gradle distributions types.
+     */
+    public enum Type {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: ❌ Should not be public, as previously discussed. What would be the purpose of making this public?

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+         */
+        VERSION
+    }
+
+    /**
+     * Returns the distribution type.
+     *
+     * @return the distribution type
+     */
+    public abstract Type getDistributionType(); // TODO (donat) rename to getType() and merge public and internal enums
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: ❌ Should not be public.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+     * <p>
+     * For {@link GradleDistributionType#VERSION} the version is returned. For
+     * {@link GradleDistributionType#LOCAL_INSTALLATION} the installation location path is returned.
+     * for {@link GradleDistributionType#REMOTE_DISTRIBUTION} the remote URI is returned. For
+     * {@link GradleDistributionType#WRAPPER} and for {@link GradleDistributionType#INVALID} an
+     * empty string is returned.
+     *
+     * @return the configuration string
+     */
+    public abstract String getConfiguration();
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: ❌ Should not be public

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import java.io.File;
+import java.net.URI;
+
+/**
+ * Represents a Gradle distribution.
+ *
+ * @author Donat Csikos
+ * @since 3.0
+ */
+public abstract class GradleDistribution {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: ❌ Should be marked as `@noimplement`. (this applies to other interfaces we have too).

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java (hunk 3 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: Removed.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java (hunk 4 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: Added `@noimplement` to thge public types.

**Comment by donat** [comment]
All review items have been resolved.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import java.net.URI;
+
+/**
+ * Represents a Gradle distribution.
+ *
+ * @author Donat Csikos
+ * @since 3.0
+ * @noimplement this interface is not intended to be implemented by clients
+ */
+public abstract class GradleDistribution {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleDistribution.java]: We should have another chat about this class on Monday. Right now it is basically useless for our end users, so we might es well hide it completely (like ProjectConnection does).

On the other hand, if we want to allow users to build their own configuration UI for instance, we should fully expose the different subtypes.

**Pull request merged**